### ISSUE TITLE
akka-http waits indefinitely for data after sending '100 Continue' when `Content-Length` header is missing or 0

### ISSUE THREAD

**Issue by ProChirathF** [playframework/playframework#10733]
When two consecutive request(GET or POST) are made with the header Expect:100-continue set, second request receives no response and a idle timeout occurs, First request is serviced immediately, but the second request is not passed to the controller or responded, blocking the client and any intermediate reverse proxy. Idle timeout setting play.server.http.idleTimeout can be used to control Playtimeout, by default this is 75 seconds.

Versions. 2.7 & 2.8 - looks like only akka http backends are affected by this issue, 2.5 with Netty backend is not affected. 

It is very straightforward to reproduce this issue, you can even use postman with the Play sample project - play-samples-play-java-hello-world-tutorial

Request 1
------------
Request Headers
```
Expect: 100-continue
User-Agent: PostmanRuntime/7.26.8
Accept: */*
Postman-Token: 11acd3d2-84ff-4967-a5ce-05d90f15e63f
Host: localhost:9000
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
```

Response Headers
```
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Permitted-Cross-Domain-Policies: master-only
Date: Mon, 01 Mar 2021 06:03:15 GMT
Content-Type: text/html; charset=UTF-8
```

Request 2 - Timeout after 75 seconds
-----------------------------------------
```
GET http://localhost:9000/
Error: read ECONNRESET
```

Request Headers
```
Expect: 100-continue
User-Agent: PostmanRuntime/7.26.8
Accept: */*
Postman-Token: 8e453fdb-559b-4c25-a22c-3f2665f63714
Host: localhost:9000
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
```

After upgrade from v2.5 we are experiencing significant issues with our integrations, it true the request is not proper, but the server should respond. 

I am also interested in any short term fixes we can apply at an application level to mitigate the issue

**Comment by mkurz** [comment]
@ProChirathF Does this happen in prod mode as well?

**Comment by mkurz** [comment]
Maybe relevant:
- Tests for 100-continue: https://github.com/playframework/playframework/blob/2.8.7/core/play-integration-test/src/it/scala/play/it/http/Expect100ContinueSpec.scala
- Some special handling from akka-http: https://github.com/playframework/playframework/blob/2.8.7/transport/server/play-akka-http-server/src/main/scala/play/core/server/AkkaHttpServer.scala#L403-L411
- https://github.com/playframework/playframework/blob/2.8.7/core/play-integration-test/src/it/scala/play/it/http/BasicHttpClient.scala#L158

**Comment by ProChirathF** [comment]
@mkurz 
Issue happens in both prod and dev.

**Comment by mkurz** [comment]
@ProChirathF Can you please provide a small reproducer project with some curl commands? Thanks!

**Comment by ProChirathF** [comment]
Sample Project

`https://github.com/playframework/play-samples/tree/2.8.x/play-java-hello-world-tutorial`

Curl call
`curl -i -H "Connection: keep-alive" -H "Expect: 100-continue" -X GET http://localhost:9000 -H "Connection: keep-alive" -H "Expect: 100-continue" -X GET http://localhost:9000`

Testing but more, it looks like any subsequent request after an Expect: 100-continue" request is not responded, second request doesn't need to be a Expect: 100-continue" request

`curl -i -H "Connection: keep-alive" -H "Expect: 100-continue" -X GET http://localhost:9000 -H "Connection: keep-alive"  -X GET http://localhost:9000`

Let me know if you any further details

**Comment by ProChirathF** [comment]
@mkurz are there any known workaround for the issue, even if it means app code change?
I tried implementing a custom HttpRequestHandler to strip-off this header, and let the route handle these request as non-expect request, it seems server handles the request at a higher level and custom HttpRequestHandler is ineffective.

https://github.com/playframework/playframework/blob/814f0c73f86eb0e85bcae7f2167c73a08fed9fd7/transport/server/play-akka-http-server/src/main/scala/play/core/server/AkkaHttpServer.scala#L402

Any advice appreciated.

**Comment by mkurz** [comment]
@ProChirathF I have no idea yet what is going on. I hope I find time to look into this issue soon.

**Comment by mkurz** [comment]
The core problem here is that akka-http incorrectly handles a `0` or missing `Content-Length` header. More details and a fix can be found here: https://github.com/akka/akka-http/pull/3786

>  are there any known workaround for the issue, even if it means app code change?

I don't think so, however this issue is only a problem if you do not send a body. As long as you send a body and the corresponding `Content-Length` header that is > 0, you should be fine.

**Comment by ProChirathF** [comment]
@ mkurz Thank you very much for the for the update. 
We are running 2.7.9 in productions, can we please have the fix in next 2.7 milestone 2.7.10 as well?

**Comment by ProChirathF** [comment]
I can reproduce the issue for POST even when the content length set correctly, 1st request is serviced, 2nd request hangs

Request 1
```
Request Headers
Expect: 100-continue
User-Agent: PostmanRuntime/7.26.10
Accept: */*
Postman-Token: 0fcc2ab2-8d5c-4fa3-8b9a-ea18707f3f95
Host: localhost:9000
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
Content-Type: application/x-www-form-urlencoded
Content-Length: 21

Request Body
password: "a"
username: "b"
```
Response 1
```
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Permitted-Cross-Domain-Policies: master-only
Date: Tue, 16 Mar 2021 06:25:26 GMT
Content-Type: text/html; charset=UTF-8
```

Request 2  - Hangs
```
Request Headers
Expect: 100-continue
User-Agent: PostmanRuntime/7.26.10
Accept: */*
Postman-Token: 0fcc2ab2-8d5c-4fa3-8b9a-ea18707f3f95
Host: localhost:9000
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
Content-Type: application/x-www-form-urlencoded
Content-Length: 21

Request Body
password: "a"
username: "b"
```

**Comment by mkurz** [comment]
@ProChirathF I am pretty sure that is because you send `Expect: 100-continue` and also send the request body immediately (without waiting for the `100 Continue` response). I updated https://github.com/akka/akka-http/pull/3786 to make sure it can handle such a case when data gets send early. But to be honest it does not make much sense that you send `Expect: 100-continue` and send all data already, because there will be no data you can continue sending with after receiving `100 Continue`... Anyway, akka-http should handle that.
BTW: The `Content-Length: 21` doesn't fit the request body you posted, probably what really gets send is `password=a&username=b` which would be 21 bytes...

One more thing:
> We are running 2.7.9 in productions, can we please have the fix in next 2.7 milestone 2.7.10 as well?

Play 2.7 is end of life and it's very likely there won't be anymore 2.7 releases... You should upgrade to Play 2.8. However, if https://github.com/akka/akka-http/pull/3786 makes it into an akka-http 10.1.15 release (backport is [here](https://github.com/akka/akka-http/pull/3787)), you can use that future version in Play by adding
```
libraryDependencies += "com.typesafe.akka" %% "akka-http-core" % "10.1.15"
```
to your `build.sbt`. That way you override the akka-http version Play 2.7 ships with.

**Comment by ProChirathF** [comment]
Thanks, agree it is a not a valid use of Expect: 100-continue to send the body immediately, however as you mentioned akka-http should handle this.

Regarding the the sample request,  the actual payload is password=a&username=b, I just copied whole request from Postman console which had prettified it.

We'll keep an eye for akka-http 10.1.15

Appreciate all the work and great communication!!

**Comment by mkurz** [comment]
Fixed by #11019
